### PR TITLE
Widget Sync - Add new widget-sync event

### LINKED ISSUE #679: Clients get out of sync from centralised datastore
### Current Behavior

If you have two separate Dashboard clients open and connected to a centralised Node-RED instance with Dashboard 2.0, the server has a single source of truth.

If one client updates that data, e.g. through a toggle switch, the centralised store of data is updated. However, other clients are not aware of this change _until_ they  refresh, which is not the desired behaviour. Any clients connected should update if a client has been the driver of change (i.e. user clicks a toggle).

This already works for all instance where a node is injected with a new value.

### Expected Behavior

All clients should reflect the server-side datastore at all times, unless using multi-user authentication/Dashboards architecture

### Steps To Reproduce

1. Add a toggle switch
2. Open two browsers connecting to the Dashboard
3. Toggle the switch on one of the clients
4. Note that the other client doesn't automically update

### Environment

- Dashboard version: 1.4.0

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### PR DESCRIPTION
**Pull request by joepavitt**
## Description

- Adds new `widget-sync` event
- Updates Events Architecture & in-app helper text to detail the new functionality
- The new event ensure that multiple clients all reflect the single source of truth, without a need to refresh the page
- This functionality is locked behind the same `validClient` test that the `msg._client` data is, so that functionality is now a consistent "single source of truth" vs. "separate clients" toggle.

### Notes

I will be following this up with #1294 as it'll clean the custom `onSync` functions out entirely, but was no need to merge it all together in the name of iteration.

## Related Issue(s)

Closes #679

### CODE REVIEW

File: ui/src/widgets/data-tracker.mjs [hunk 1]
```diff
@@ -68,6 +68,21 @@ export function useDataTracker (widgetId, onInput, onLoad, onDynamicProperties)
...
+            onSync(msg)
+        } else {
+            // only need the msg.payload
+            store.commit('data/bind', {
+                widgetId,
+                msg: {
+                    payload: msg.payload
+                }
+            })
+        }
```
[the remark this answers is not in the document]
Comment (joepavitt) [reply ui/src/widgets/data-tracker.mjs:83]: We should be able to get the knowledge of the mapping from the widget properties at this point and map accordingly. Fiddly, but possible.

File: ui/src/widgets/ui-number-input/UINumberInput.vue [hunk 2]
```diff
@@ -196,6 +196,12 @@ export default {
                 this.previousValue = msg.payload
             }
         },
+        onSync (msg) {
+            if (msg?.payload !== undefined) {
```
[the remark this answers is not in the document]
Comment (joepavitt) [reply ui/src/widgets/ui-number-input/UINumberInput.vue:200]: This still needs the `?` guard on it in the case of an `undefined` instance of `msg`

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply ui/src/widgets/ui-number-input/UINumberInput.vue:200]: indeed, this one was an overzealous (I could have sworn I seen `if (msg) { ` immediately above it like the others!

Time for a vacation :D

File: docs/contributing/guides/events.md [hunk 3]
```diff
@@ -54,6 +54,14 @@ Some examples of events that are emitted from the Dashboard to Node-RED include:
 - `widget-action` - When a user interacts with a widget, and state of the widget is not important, e.g. a button click
 - `widget-send` - Used by `ui-template` to send a custom `msg` object, e.g. `send(msg)`, which will be stored in the server-side data store.
 
+#### Syncing Widgets
+
+The `widget-change` event is used to emit input from a ser, and represents a change of state for that widget, e.g. a switch can be on/off by a user clicking it. In this case, if you have multiple clients connected to the same Node-RED instance, Dashboard will ensure that clients are in-sync when values change.
```
[the remark this answers is not in the document]
Comment (joepavitt) [reply docs/contributing/guides/events.md]: Resolved in https://github.com/FlowFuse/node-red-dashboard/pull/1463/commits/2f95a3bc1d94d39efba9f1ead622178585d64446

**Pull request merged**